File: Cpp/Mnist.cpp
#include "Mnist.h"

size_t GetFileSize(std::wstring path)
{
	std::ifstream in(path, std::ifstream::ate | std::ifstream::binary);
	auto rtn = in.tellg();
	in.close();
	return rtn;
}

MnistItem::MnistItem()
	: _Image(nullptr), _Label(0), _Rows(0), _Columns(0)
{
}

MnistItem::MnistItem(void * data, unsigned char label, int rows, int columns)
	: _Label(label), _Image(new unsigned char[rows * columns]), _Rows(rows), _Columns(columns)
{
	memcpy(_Image, data, rows * columns);
}

MnistItem::MnistItem(const MnistItem & other)
	: _Label(other._Label), _Image(new unsigned char[other._Rows * other._Columns]), _Rows(other._Rows), _Columns(other._Columns)
{
	memcpy(_Image, other._Image, _Rows * _Columns);
}

MnistItem::MnistItem(MnistItem && other)
	: _Label(other._Label), _Image(other._Image), _Rows(other._Rows), _Columns(other._Columns)
{
	other._Image = nullptr;
}

MnistItem & MnistItem::operator=(const MnistItem & other)
{
	_Label = other._Label;
	_Image = new unsigned char[other._Rows * other._Columns];
	_Rows = other._Rows;
	_Columns = other._Columns;
	memcpy(_Image, other._Image, _Rows * _Columns);
	return *this;
}

MnistItem & MnistItem::operator=(MnistItem && other)
{
	_Label = other._Label;
	_Image = other._Image;
	_Rows = other._Rows;
	_Columns = other._Columns;
	other._Image = nullptr;
	return *this;
}

MnistItem::~MnistItem()
{
	delete[] _Image;
}

int Mnist::ByteToInt(unsigned char * num)
{
	int result = 0;
	int offset[] = { 24, 16, 8, 0 };
	for (int i = 0; i < 4; i++)
	{
		result |= ((int)num[i]) << offset[i];
	}
	return result;
}

Mnist::Mnist(std::wstring imagePath, std::wstring labelPath, bool normalize)
{
	std::ifstream imageRead(imagePath, std::ifstream::binary);
	std::ifstream labelRead(labelPath, std::ifstream::binary);
	unsigned char * imageTemp = new unsigned char[GetFileSize(imagePath)];
	unsigned char * labelTemp = new unsigned char[GetFileSize(labelPath)];
	imageRead.read((char *)imageTemp, GetFileSize(imagePath));
	labelRead.read((char *)labelTemp, GetFileSize(labelPath));
	int imageMagicNumber = ByteToInt(imageTemp + 0);
	int labelMagicNumber = ByteToInt(labelTemp + 0);
	if (imageMagicNumber != 2051)
	{
		delete[] imageTemp;
		delete[] labelTemp;
		throw std::exception("Not a valid MNIST image data file.");
	}

	if (labelMagicNumber != 2049)
	{
		delete[] imageTemp;
		delete[] labelTemp;
		throw std::exception("Not a valid MNIST label data file.");
	}

	int imageLength = ByteToInt(imageTemp + 4);
	int labelLength = ByteToInt(labelTemp + 4);
	if (imageLength != labelLength)
	{
		delete[] imageTemp;
		delete[] labelTemp;
		throw std::exception("Number of items of two files are not the same.");
	}

	int imageRows = ByteToInt(imageTemp + 8);
	int imageColumns = ByteToInt(imageTemp + 12);

	_Length = imageLength;
	if (normalize)
	{
		_Rows = 1;
		_Columns = imageRows * imageColumns;
	}
	else
	{
		_Rows = imageRows;
		_Columns = imageColumns;
	}
	_Image = new MnistItem[_Length];

	for (int i = 0; i < _Length; i++)
	{
		_Image[i] = MnistItem(imageTemp + 16 + i * _Rows * _Columns, labelTemp[8 + i], _Rows, _Columns);
	}

	delete[] imageTemp;
	delete[] labelTemp;
}

Mnist::~Mnist()
{
	delete[] _Image;
}

File: CS/Mnist.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CntkMnistPractice
{
    class Mnist : IEnumerable<MnistItem>
    {
        MnistItem[] image;

        // Returns the number of items
        public int Length { get; }
        // Returns the number of rows of an image.
        public int Rows { get; }
        // Returns the number of columns of an image.
        public int Columns { get; }

        public MnistItem GetAt(int i)
        {
            return image[i];
        }

        private static int ReverseEndian(int num)
        {
            int[] filter = new int[4] { 0x7F000000, 0x00FF0000, 0x0000FF00, 0x000000FF };
            int[] offset = new int[4] { 24, 8, 8, 24 };
            
            int result = 0;
            for (int i = 0; i < 2; i++)
                result |= (num & filter[i]) >> offset[i];
            for (int i = 2; i < 4; i++)
                result |= (num & filter[i]) << offset[i];
            if (num < 0)
                result += 0x00000080;

            return result;
        }

        public Mnist(string imagePath, string labelPath, bool normalize = false)
        {
            byte[] imageTemp = System.IO.File.ReadAllBytes(imagePath);
            byte[] labelTemp = System.IO.File.ReadAllBytes(labelPath);
            int imageMagicNumber = ReverseEndian(BitConverter.ToInt32(imageTemp, 0));
            int labelMagicNumber = ReverseEndian(BitConverter.ToInt32(labelTemp, 0));
            if (imageMagicNumber != 2051)
                throw new Exception("Not a valid MNIST image data file.");

            if (labelMagicNumber != 2049)
                throw new Exception("Not a valid MNIST label data file.");

            int imageLength = ReverseEndian(BitConverter.ToInt32(imageTemp, 4));
            int labelLength = ReverseEndian(BitConverter.ToInt32(labelTemp, 4));
            if (imageLength != labelLength)
                throw new Exception("Number of items of two files are not the same.");

            int imageRows = ReverseEndian(BitConverter.ToInt32(imageTemp, 8));
            int imageColumns = ReverseEndian(BitConverter.ToInt32(imageTemp, 12));

            Length = imageLength;
            if(normalize)
            {
                Rows = 1;
                Columns = imageRows * imageColumns;
            }
            else
            {
                Rows = imageRows;
                Columns = imageColumns;
            }
            image = new MnistItem[Length];
            
            for (int i = 0; i < Length; i++)
            {
                byte[] item = new byte[Rows * Columns];
                Array.Copy(imageTemp, 16 + i * Rows * Columns, item, 0, Rows * Columns);
                image[i] = new MnistItem(item, labelTemp[8 + i], Rows, Columns);
            }

        }
        
        public IEnumerator<MnistItem> GetEnumerator()
        {
            for (int i = 0; i < Length; i++)
            {
                yield return image[i];
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return (IEnumerator)GetEnumerator();
        }
    }

    class MnistItem
    {
        public byte[] Image { get; }
        public byte Label { get; }

        public int Rows { get; }
        public int Columns { get; }

        public MnistItem(byte[] data, byte label, int rows, int columns)
        {
            this.Label = label;
            this.Image = new byte[rows * columns];
            this.Rows = rows;
            this.Columns = columns;
            data.CopyTo(this.Image, 0);
        }
    }

    class NormalizedMnistItem
    {
        public float[] Image { get; }
        public float[] Label { get; }
        public int Rows { get; }
        public int Columns { get; }

        public NormalizedMnistItem(MnistItem item)
        {
            Rows = item.Rows;
            Columns = item.Columns;
            Image = new float[Rows * Columns];
            Label = new float[10];
            Label[item.Label] = 1.0f;
            for (int i = 0; i < item.Image.Length; i++)
            {
                Image[i] = ((float)item.Image[i]) / byte.MaxValue;
            }
        }
    }

    class NormalizedMnistItemDouble
    {
        public double[] Image { get; }
        public double[] Label { get; }
        public int Rows { get; }
        public int Columns { get; }

        public NormalizedMnistItemDouble(MnistItem item)
        {
            Rows = item.Rows;
            Columns = item.Columns;
            Image = new double[Rows * Columns];
            Label = new double[10];
            Label[item.Label - 1] = 1.0f;
            for (int i = 0; i < item.Image.Length; i++)
            {
                Image[i] = ((double)item.Image[i]) / byte.MaxValue;
            }
        }
    }
}
